vo_gpu: revert 8a09299 and conditionally clear framebuffer by default

Akemi opened the pull request:
Fixes #8273

Dudemanguy commented:
Just leaving my 2 cents. On x11 with amdgpu and opengl, I always see the garbled frames mentioned in that commit. Conditionally or always clearing the framebuffer seems to make no difference. With or without compositing on, switching fullscreen, resizing, etc. all leads to weird artifacts. This is with linux 5.9.10 and mesa 20.2.3 so who knows what changed since that commit was originally made.

Of course someone else may have a different findings but right now I don't see why we can't just revert it. There doesn't seem to be any benefit on my end from always clearing the framebuffer. @aufkrawall does the current mpv master behavior of always clearing the framebuffer still have improvements for you over conditionally clearing it?

aufkrawall commented:
It hasn't changed since my last comment on it, 8a0929973de15d9574595c5a098bb3446757ef16 didn't help much. I still use `radeonsi_zerovram` in `.drirc`, which gets rid of it entirely (I think the env var for it has been removed recently). Imho a MR for Mesa, so that it does this by default, still seems like an attractive thing to do.

Akemi commented:
reopening this to start another discussion since the changes of the other accepted PR (#8372) were reverted again (d9008d2aa84467064c3d76407f5f6a5b86e76520).

i didn't change anything yet to safe me from unnecessary work, eg i would like to first sort out some question. is this acceptable? if not, is anything acceptable and how should it be done? if the changes here are acceptable, what should be the default (since i guess the default should be always) and/or what should be changed?

what's your opinion on that @decent-name? i don't want to play revert ping pong with you.

decent-name commented:
If this is really needed, you can probably make it conditional. It's not the first case. For the most part, I want it to work on x11 without me having to lift a finger.

Akemi commented:
okay if you are fine with the default like this (conditional) i am leaving it as is. also saw it doesn't work in dumb mode so i added it to the whitelist and rebased with current master.

if you are fine with the changes like this, it would be nice if you could approve the PR or merge it.

decent-name commented on video/out/gpu/video.c:
```diff
@@ -448,6 +448,8 @@ const struct m_sub_options gl_video_conf = {
         {"deband", OPT_SUBSTRUCT(deband_opts, deband_conf)},
         {"sharpen", OPT_FLOAT(unsharp)},
+        {"clear-framebuffer", OPT_CHOICE(clear_framebuffer,
+            {"conditional", 0}, {"always", 1}, {"never", -1})},
```
"always" should be the default then.

decent-name replied:
Also, "never" actually makes no sense, because it'll display old framebuffer contents.

Akemi replied:
my intend was to make it easier for me to debug user problems related to rendering or that specific bug. i could just tell them to test with 'never' and if it fixes their issue i can just direct them upstream to apple. telling them to build mpv themselves and commenting out the clearing code is usually too hard for most of them.

if this is not enough for it to exist, fine by me?

Akemi replied:
int he case of removal i would just make a OPT_BOOL out of it

decent-name commented on video/out/gpu/video.c:
```diff
@@ -3251,9 +3253,13 @@ void gl_video_render_frame(struct gl_video *p, struct vo_frame *frame,
-    struct m_color c = p->clear_color;
-    float clear_color[4] = {c.r / 255.0, c.g / 255.0, c.b / 255.0, c.a / 255.0};
-    p->ra->fns->clear(p->ra, fbo.tex, clear_color, &target_rc);
+    if (p->opts.clear_framebuffer == 1 || p->opts.clear_framebuffer == 0 &&
```
Needs ( ).

Akemi replied:
just a question on the side, is that some code convention we have and i missed? since technically it works as intended.

decent-name replied:
When mixing && and || (and with some other operators). Not needed when it's a serious of ||. It's not explicitly in the coding style, but most code does it.

Akemi replied:
okay. i think explicitly using brackets here makes it easier to read, since it's explicit. at some point i added them but removed them for the commit again, since i wasn't sure what would be preferred.

Akemi commented:
i made the changes, updated the docs and the commit message.
